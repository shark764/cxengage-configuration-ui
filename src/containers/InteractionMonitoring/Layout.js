import React, { Component } from 'react';
import PropTypes from 'prop-types';
import styled, { injectGlobal } from 'styled-components';
import 'rxjs/add/operator/map';
import {
  messageSubscribe,
  messageUnsubscribe
} from './windowMessageObservable';
import {
  localStorageSubscribe,
  localStorageUnsubscribe
} from './localStorageObservable';
import CheckboxFilterMenu from '../CheckboxFilterMenu';
import { Button, PageHeader, InteractionDetails } from 'cx-ui-components';
import ReactTable from 'react-table';
import interactionIdColumn from '../../components/tableColumns/interactionId';
import expanderColumn from '../../components/tableColumns/expander';
import agentColumn from '../../components/tableColumns/agent';
import customerColumn from '../../components/tableColumns/customer';
import contactPointColumn from '../../components/tableColumns/contactPoint';
import flowColumn from '../../components/tableColumns/flow';
import channelColumn from '../../components/tableColumns/channel';
import directionColumn from '../../components/tableColumns/direction';
import presenceStateColumn from '../../components/tableColumns/presence';
import startDateColumn from '../../components/tableColumns/startDate';
import startTimeColumn from '../../components/tableColumns/startTime';
import elapsedTimeColumn from '../../components/tableColumns/elapsedTime';
import monitoringColumn from '../../components/tableColumns/monitoring';
import groupsColumn from '../../components/tableColumns/groups';
import skillsColumn from '../../components/tableColumns/skills';

const InlineCheckboxFilterMenu = styled(CheckboxFilterMenu)`
  display: inline-block;
`;
const ConverTimeButton = styled(Button)`
  margin-right: 20px;
`;

const Wrapper = styled.div`
  padding: 20px;
`;

injectGlobal`
.InteractionMonitoringTable .rt-tbody {
  overflow: initial;
}
`;

export default class InteractionMonitoring extends Component {
  constructor() {
    super();
    this.state = {
      filtered: [{ id: 'channel', value: 'voice' }]
    };
  }

  componentWillMount() {
    messageSubscribe();
    localStorageSubscribe();
    this.props.setCurrentEntity('InteractionMonitoring');
    this.props.fetchData('groups', 'InteractionMonitoring');
    this.props.fetchData('skills', 'InteractionMonitoring');
    this.props.startInteractionMonitoring();
  }

  componentWillUnmount() {
    messageUnsubscribe();
    localStorageUnsubscribe();
  }
  defaultFilterMethod = ({ value, id }, row) =>
    String(row[id])
      .toLowerCase()
      .indexOf(value.toLowerCase()) > -1;

  getTableRowProps = (state, rowInfo) => {
    if (rowInfo) {
      return {
        onClick: () => {
          if (this.props.selected === rowInfo.row.interactionId) {
            return this.props.removeSelected();
          } else {
            return this.props.setSelected(rowInfo.row.interactionId, {
              [rowInfo.viewIndex]: rowInfo.row.interactionId
            });
          }
        },
        style: {
          background:
            rowInfo.row.interactionId === this.props.monitoredId
              ? 'rgba(253, 255, 50, 0.17)'
              : null
        }
      };
    } else {
      return { style: {} };
    }
  };
  getTableProps = () => ({ style: { height: '80vh' } });
  getTdProps = () => ({ style: { fontSize: '11.5pt' } });
  getTheadProps = () => ({ style: { color: 'grey' } });
  onFilteredChange = filtered => this.setState({ filtered: filtered });
  onSortedChange = sorted => this.props.setSorted(sorted);
  SubComponent = ({
    row: {
      startTimestamp,
      agentName,
      direction,
      channel,
      contactPoint,
      flowName,
      customer,
      monitoring
    }
  }) => (
    <InteractionDetails
      data={{
        startTimestamp,
        agentName,
        direction,
        channel,
        contactPoint,
        flowName,
        customer,
        monitoring
      }}
      twelveHourFormat={this.props.twelveHourFormat}
    />
  );
  toggleTimeFormat = () => this.props.toggleTimeFormat();

  render() {
    return (
      <Wrapper>
        <PageHeader
          text="Interaction Monitoring"
          helpLink="https://docs.cxengage.net/Help/Content/Monitoring/Silent%20Monitoring/Silent-monitoring.htm?Highlight=silent%20monitoring"
        >
          <div>
            <ConverTimeButton
              id="timeConversion"
              type="secondary"
              className="timeConversion"
              onClick={this.toggleTimeFormat}
            >
              {this.props.twelveHourFormat ? '12h' : '24h'}
            </ConverTimeButton>
            <InlineCheckboxFilterMenu
              type="secondary"
              menuType="Columns"
              tableType="InteractionMonitoring"
              selectionType="checkbox"
            >
              Columns
            </InlineCheckboxFilterMenu>
          </div>
        </PageHeader>

        {!this.props.areAllColNotActive && (
          <ReactTable
            defaultPageSize={20}
            className="InteractionMonitoringTable -striped -highlight"
            data={this.props.tableData}
            filterable
            defaultFilterMethod={this.defaultFilterMethod}
            expanded={this.props.expanded}
            sorted={this.props.sorted}
            filtered={this.state.filtered}
            onFilteredChange={this.onFilteredChange}
            onSortedChange={this.onSortedChange}
            getTableProps={this.getTableProps}
            getTdProps={this.getTdProps}
            getTheadProps={this.getTheadProps}
            getTrProps={this.getTableRowProps}
            SubComponent={this.SubComponent}
            columns={[
              expanderColumn(),
              interactionIdColumn(this.props.activeColumns[0]),
              agentColumn(this.props.activeColumns[1]),
              customerColumn(this.props.activeColumns[2]),
              contactPointColumn(this.props.activeColumns[3]),
              flowColumn(this.props.activeColumns[4]),
              channelColumn(this.props.activeColumns[5]),
              directionColumn(
                this.props.activeColumns[6],
                'InteractionMonitoring'
              ),
              presenceStateColumn(this.props.activeColumns[7]),
              startDateColumn(
                this.props.activeColumns[8],
                'InteractionMonitoring'
              ),
              startTimeColumn(
                this.props.activeColumns[9],
                'InteractionMonitoring',
                this.props.twelveHourFormat
              ),
              elapsedTimeColumn(
                this.props.activeColumns[10],
                'InteractionMonitoring'
              ),
              monitoringColumn(
                this.props.activeColumns[11],
                'InteractionMonitoring',
                this.props.monitoredId,
                this.props.monitoringStatus
              ),
              groupsColumn(
                this.props.activeColumns[12],
                'InteractionMonitoring'
              ),
              skillsColumn(
                this.props.activeColumns[13],
                'InteractionMonitoring'
              )
            ]}
          />
        )}
      </Wrapper>
    );
  }
}

InteractionMonitoring.propTypes = {
  monitoredId: PropTypes.string,
  monitoringStatus: PropTypes.string,
  totalRatio: PropTypes.array,
  tableData: PropTypes.any,
  expanded: PropTypes.object.isRequired,
  selected: PropTypes.string.isRequired,
  areAllColNotActive: PropTypes.bool.isRequired,
  sorted: PropTypes.any.isRequired,
  toggleTimeFormat: PropTypes.func.isRequired,
  updateTableData: PropTypes.func.isRequired,
  updateSkillsColumnFilter: PropTypes.func.isRequired,
  updateGroupsColumnFilter: PropTypes.func.isRequired,
  setExpanded: PropTypes.func.isRequired,
  setSelected: PropTypes.func.isRequired,
  setSorted: PropTypes.func.isRequired,
  removeSelected: PropTypes.func.isRequired,
  twelveHourFormat: PropTypes.bool.isRequired,
  activeColumns: PropTypes.arrayOf(PropTypes.bool.isRequired),
  setCurrentEntity: PropTypes.func,
  fetchData: PropTypes.func,
  startInteractionMonitoring: PropTypes.func
};
