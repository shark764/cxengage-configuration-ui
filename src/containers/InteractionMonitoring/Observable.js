import { fromEvent } from 'rxjs/observable/fromEvent';
import 'rxjs/add/operator/map';
import 'rxjs/add/operator/filter';
import { updateTableData } from '../../redux/modules/reporting/interactionMonitoring';

import store from '../../redux/store.js';

export function messageObservable() {
  return fromEvent(window, 'message')
    .filter(
      ({ data }) =>
        data.subscription &&
        data.subscription.topic === 'cxengage/reporting/batch-response'
    )
    .map(event =>
      store.dispatch(
        updateTableData(
          event.data.subscription.response.interactionsInConversationList.body
            .results.interactions
        )
      )
    );
}

let subscription;
export function subscribe() {
  subscription = messageObservable().subscribe();
}
export function unsubscribe() {
  subscription.unsubscribe();
}
