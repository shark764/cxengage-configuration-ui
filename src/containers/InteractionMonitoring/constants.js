export const SET_TABLE_DATA = 'src/SupervisorToolbar/SET_TABLE_DATA';
export const SET_SORTED = 'src/SupervisorToolbar/SET_SORTED';
export const SET_FILTERED = 'src/SupervisorToolbar/SET_FILTERED';
export const SET_EXPANDED = 'src/SupervisorToolbar/SET_EXPANDED';
export const SET_SELECTED = 'src/SupervisorToolbar/SET_SELECTED';
export const REMOVE_SELECTED = 'src/SupervisorToolbar/REMOVE_SELECTED';
