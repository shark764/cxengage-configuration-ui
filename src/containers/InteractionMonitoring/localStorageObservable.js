import { fromJS } from 'immutable';
import { fromEvent } from 'rxjs/observable/fromEvent';
import 'rxjs/add/operator/map';
import { updateAllOfSupervisorToolbar } from '../../redux/modules/supervisorToolbar';
import store from '../../redux/store.js';

export function localStorageObservable() {
  return fromEvent(window, 'storage').map(response =>
    store.dispatch(
      updateAllOfSupervisorToolbar(
        fromJS(JSON.parse(localStorage.getItem('SupervisorToolbar')))
      )
    )
  );
}

let subscription;
export function localStorageSubscribe() {
  subscription = localStorageObservable().subscribe();
}
export function localStorageUnsubscribe() {
  subscription.unsubscribe();
}
