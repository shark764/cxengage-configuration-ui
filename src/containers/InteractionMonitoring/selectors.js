import { createSelector } from 'reselect';

const selectInteractionMonitoringMap = state =>
  state.get('InteractionMonitoring');

export const selectInteractionMonitoring = createSelector(
  selectInteractionMonitoringMap,
  interactionMonitoring => interactionMonitoring
);

export const selectInteractionMonitoringTableData = createSelector(
  selectInteractionMonitoringMap,
  subState => {
    console.log(subState.toJS());
    return subState.toJS();
  }
  // // console.log(subState.get("tableData").size);
  // if (subState.get("data").size === 0) {
  //   //   console.log(subState.get("tableData"));
  //   return [];
  // } else {
  //   //   console.log(subState.get("tableData"));
  //   return [subState.get("data")];
  // }
  //   }
);
