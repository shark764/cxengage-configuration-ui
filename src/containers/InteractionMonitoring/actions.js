import * as ACTIONS from './constants';

export function updateTableData(arrayOfTableData) {
  return {
    type: ACTIONS.SET_TABLE_DATA,
    arrayOfTableData
  };
}
export function setExpanded(expanded) {
  return {
    type: ACTIONS.SET_EXPANDED,
    expanded
  };
}
export function setFiltered(filtered) {
  return {
    type: ACTIONS.SET_FILTERED,
    filtered
  };
}
export function setSelected(selected, expanded) {
  return {
    type: ACTIONS.SET_SELECTED,
    selected,
    expanded
  };
}
export function setSorted(sorted) {
  return {
    type: ACTIONS.SET_SORTED,
    sorted
  };
}
export function removeSelected() {
  return {
    type: ACTIONS.REMOVE_SELECTED
  };
}
