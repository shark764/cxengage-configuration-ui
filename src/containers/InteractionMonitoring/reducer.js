import * as ACTIONS from './constants';
import { fromJS } from 'immutable';

const initialState = fromJS({
  data: [],
  sorted: [],
  filtered: [],
  expanded: {},
  selected: ''
});

function InteractionMonitoring(state = initialState, action) {
  switch (action.type) {
    case ACTIONS.SET_TABLE_DATA:
      return state.set('data', fromJS(action.arrayOfTableData));
    case ACTIONS.SET_SORTED:
      return state
        .set('sorted', action.sorted)
        .set('selected', '')
        .set('expanded', {});
    case ACTIONS.SET_FILTERED:
      return state
        .set('filtered', action.filtered)
        .set('expanded', {})
        .set('selected', '');
    case ACTIONS.SET_EXPANDED:
      return state.set('expanded', action.expanded);
    case ACTIONS.SET_SELECTED:
      return state
        .set('selected', action.selected)
        .set('expanded', action.expanded);
    case ACTIONS.REMOVE_SELECTED:
      return state.set('expanded', {}).set('selected', '');
    default:
      return state;
  }
}

export default InteractionMonitoring;
