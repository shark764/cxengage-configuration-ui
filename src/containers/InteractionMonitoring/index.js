import { connect } from 'react-redux';
import {
  selectInteractionMonitoringActiveColumns,
  areAllColNotActive,
  selectTimeFormat,
  totalRatio,
  activeMenuItems,
  areAllActive
} from '../../redux/modules/columnFilterMenus/selectors';
import { setCurrentEntity, fetchData } from '../../redux/modules/entities';
import {
  updateTableData,
  setExpanded,
  setSorted,
  setSelected,
  removeSelected,
  startInteractionMonitoring,
  getExtensions
} from '../../redux/modules/reporting/interactionMonitoring';
import {
  updateSkillsColumnFilter,
  updateGroupsColumnFilter,
  toggleTimeFormat
} from '../../redux/modules/columnFilterMenus';
import {
  selectInteractionMonitoringTableData,
  selectInteractionMonitoringSorted,
  selectInteractionMonitoringExpanded,
  selectInteractionMonitoringSelected,
  userHasViewAllMonitoredCallsPermission,
  userHasMonitorAllCallsPermission,
  selectExtensions,
  isFetchingUserExtensions
} from '../../redux/modules/reporting/interactionMonitoring/selectors';
import {
  selectSupervisorToolbarSilentMonitoringInteractionId,
  selectSupervisorToolbarSilentMonitoringStatus,
  selectCanSilentMonitor,
  selectLoadingUserStatus
} from '../../redux/modules/supervisorToolbar/selectors';
import { getCurrentAgentId } from '../../redux/modules/userData/selectors';
import { getHelpLink } from '../EntityTable/selectors';
import { entitiesMetaData } from '../../redux/modules/entities/metaData';

import Layout from './layout';
import { isInIframe } from 'serenova-js-utils/browser';

const TABLE_TYPE = 'interactionMonitoring';

export const mapStateToProps = (state, props) => ({
  pageTitle: entitiesMetaData[TABLE_TYPE] ? entitiesMetaData[TABLE_TYPE].pageTitle : '',
  pageHelpLink: getHelpLink(state),
  getCurrentAgentId: getCurrentAgentId(state),
  areAllColNotActive: areAllColNotActive(state, {
    menuType: 'Columns',
    tableType: TABLE_TYPE
  }),
  activeGroupFilters: activeMenuItems(state, { tableType: TABLE_TYPE, menuType: 'Groups' }).toJS(),
  activeSkillFilters: activeMenuItems(state, { tableType: TABLE_TYPE, menuType: 'Skills' }).toJS(),
  groupsAreAllActive: areAllActive(state, { tableType: TABLE_TYPE, menuType: 'Groups' }),
  skillsAreAllActive: areAllActive(state, { tableType: TABLE_TYPE, menuType: 'Skills' }),
  extensions: selectExtensions(state),
  totalRatio: totalRatio(state, props),
  activeColumns: selectInteractionMonitoringActiveColumns(state, props),
  twelveHourFormat: selectTimeFormat(state, props),
  tableData: selectInteractionMonitoringTableData(state, props),
  expanded: selectInteractionMonitoringExpanded(state, props),
  selected: selectInteractionMonitoringSelected(state, props),
  sorted: selectInteractionMonitoringSorted(state, props),
  monitoredId: selectSupervisorToolbarSilentMonitoringInteractionId(state, props),
  monitoringStatus: selectSupervisorToolbarSilentMonitoringStatus(state, props),
  userHasViewAllMonitoredCallsPermission: userHasViewAllMonitoredCallsPermission(state),
  userHasMonitorAllCallsPermission: userHasMonitorAllCallsPermission(state),
  insideIframe: !isInIframe(),
  canSilentMonitor: selectCanSilentMonitor(state),
  loadingUserStatus: selectLoadingUserStatus(state),
  isFetchingUserExtensions: isFetchingUserExtensions(state)
});

export const actions = {
  toggleTimeFormat,
  updateTableData,
  updateSkillsColumnFilter,
  updateGroupsColumnFilter,
  setExpanded,
  setSelected,
  setSorted,
  removeSelected,
  fetchData,
  startInteractionMonitoring,
  getExtensions,
  setCurrentEntity
};

export default connect(mapStateToProps, actions)(Layout);
