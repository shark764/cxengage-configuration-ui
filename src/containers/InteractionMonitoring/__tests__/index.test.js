/*
 * Copyright Â© 2015-2018 Serenova, LLC. All rights reserved.
 */

import React from 'react';
import { createStore } from 'redux';
import { shallow } from 'enzyme';
import InteractionMonitoring, { mapStateToProps } from '../';
jest.mock('../../../redux/store.js', () => jest.fn());

// Mock all the required selectors
jest.mock('../../../redux/modules/columnFilterMenus/selectors', () => ({
  selectInteractionMonitoringActiveColumns: () => [true],
  areAllColNotActive: () => false,
  selectTimeFormat: () => true,
  totalRatio: () => []
}));
jest.mock(
  '../../../redux/modules/reporting/interactionMonitoring/selectors',
  () => ({
    selectInteractionMonitoringTableData: () => ({ item1: '', item2: '' }),
    selectInteractionMonitoringSorted: () => true,
    selectInteractionMonitoringExpanded: () => ({}),
    selectInteractionMonitoringSelected: () => 'conected',
    userHasViewAllMonitoredCallsPermission: () => true
  })
);
jest.mock('../../../redux/modules/supervisorToolbar/selectors', () => ({
  selectSupervisorToolbarSilentMonitoringInteractionId: () =>
    '0000-0000-0000-0000',
  selectSupervisorToolbarSilentMonitoringStatus: () => 'mockStatus'
}));
jest.mock('../../../redux/modules/userData/selectors', () => ({
  getCurrentAgentId: () => '0000-0000-0000-0000'
}));

describe('InteractionMonitoring Renders', () => {
  it('renders', () => {
    const store = createStore(state => state);
    shallow(<InteractionMonitoring store={store}>Child</InteractionMonitoring>);
  });
});

describe('Maps state to props only using selectors', () => {
  it('validates object created from mapStateToProps', () => {
    expect(mapStateToProps()).toMatchSnapshot();
  });
});
