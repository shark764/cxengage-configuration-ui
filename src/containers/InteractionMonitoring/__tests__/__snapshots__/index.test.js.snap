// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Maps state to props only using selectors validates object created from mapStateToProps 1`] = `
Object {
  "activeColumns": Array [
    true,
  ],
  "areAllColNotActive": false,
  "expanded": Object {},
  "getCurrentAgentId": "0000-0000-0000-0000",
  "insideIframe": false,
  "monitoredId": "0000-0000-0000-0000",
  "monitoringStatus": "mockStatus",
  "pageHelpLink": undefined,
  "pageTitle": "Interaction Monitoring",
  "selected": "conected",
  "sorted": true,
  "tableData": Object {
    "item1": "",
    "item2": "",
  },
  "totalRatio": Array [],
  "twelveHourFormat": true,
  "userHasMonitorAllCallsPermission": true,
  "userHasViewAllMonitoredCallsPermission": true,
}
`;
